Gems/Atom/RHI/Metal/Code/Source/RHI/BindlessArgumentBuffer.cpp

```diff
@@ -139,85 +139,112 @@ namespace AZ::Metal
 
     uint32_t BindlessArgumentBuffer::AttachReadImage(id <MTLTexture> mtlTexture)
     {
-        uint32_t allocatorIndex = static_cast<uint32_t>(RHI::ShaderResourceGroupData::BindlessResourceType::ReadTexture);
-        uint32_t heapIndex = static_cast<uint32_t>(m_allocators[allocatorIndex].Allocate(1, 1).m_ptr);
-        if(m_unboundedArraySupported)
-        {
-            m_bindlessTextureArgBuffer->UpdateTextureView(mtlTexture, heapIndex);
-        }
-        else
+        RHI::VirtualAddress address;
         {
-            m_boundedArgBuffer->UpdateTextureView(mtlTexture, (heapIndex+(UnboundedArraySize*allocatorIndex)));
+            AZStd::lock_guard<AZStd::mutex> lock(m_mutex);
+            uint32_t allocatorIndex = static_cast<uint32_t>(RHI::ShaderResourceGroupData::BindlessResourceType::ReadTexture);
+            address = m_allocators[allocatorIndex].Allocate(1, 1).m_ptr;
+            AZ_Assert(address.IsValid(), "Bindless allocator ran out of space.");
+            uint32_t heapIndex = static_cast<uint32_t>(address.m_ptr);
+            if (m_unboundedArraySupported)
+            {
+                m_bindlessTextureArgBuffer->UpdateTextureView(mtlTexture, heapIndex);
+            }
+            else
+            {
+                m_boundedArgBuffer->UpdateTextureView(mtlTexture, (heapIndex + (UnboundedArraySize * allocatorIndex)));
+            }
         }
-        return heapIndex;
+        return static_cast<uint32_t>(address.m_ptr);
     }
     
     uint32_t BindlessArgumentBuffer::AttachReadWriteImage(id <MTLTexture> mtlTexture)
     {
-        uint32_t allocatorIndex = static_cast<uint32_t>(RHI::ShaderResourceGroupData::BindlessResourceType::ReadWriteTexture);
-        uint32_t heapIndex = static_cast<uint32_t>(m_allocators[allocatorIndex].Allocate(1, 1).m_ptr);
-        if(m_unboundedArraySupported)
+        RHI::VirtualAddress address;
         {
-            m_bindlessRWTextureArgBuffer->UpdateTextureView(mtlTexture, heapIndex);
-        }
-        else
-        {
-            m_boundedArgBuffer->UpdateTextureView(mtlTexture, (heapIndex+(UnboundedArraySize*allocatorIndex)));
+            AZStd::lock_guard<AZStd::mutex> lock(m_mutex);
+            uint32_t allocatorIndex = static_cast<uint32_t>(RHI::ShaderResourceGroupData::BindlessResourceType::ReadWriteTexture);
+            address = static_cast<uint32_t>(m_allocators[allocatorIndex].Allocate(1, 1).m_ptr);
```

Comment:
This line want's fixing up as well as it's down casting to a uint32_t from a ptr and then contructing a new 64 bit virtual address.